### DISCUSSION TITLE
How to use your extended kalman filter on a linux computer?

### DISCUSSION THREAD

**Discussion by serkanMzlm** [bitcraze/discussions#395] [Q&A]
We want to build a quadcopter with using a linux computer jetson nano etc. We need to filter data from sensors (MPU9250, VL53L1x, PMW3901).  For this we want to use your extended kalman filter. We looked at your software and we didn't figure out how to use .
Firstly is it possible to use this code for our hardware? Because you are using STM32 microcontroller.

If it is possible to use, how can we modify this code for our purpose?

Sorry we are new to this stuff, so your helps is so precious to us.

**Comment by krichardsson** [comment]
> Firstly is it possible to use this code for our hardware? Because you are using STM32 microcontroller.

Unfortunately I don't think it will be straight forward, it has not really been designed with portability in mind. I tried to break out the core to reduce the dependencies to FreeRTOS but I think it still uses matrix operations that are not platform independent. 

Code in src/utils is more portable, but the estimators are unfortunately not in there.

**Comment by serkanMzlm** [comment]
We convert RTOS codes to ROS2. We also changed arm's functions and structs so we have ability to use your codes. We understand that updateQueuedMeasurements() function update coredata struct. There is a switch control idiom. How this idiom work? Which case is necesssary for position controller? Our main purpose is prepare inputs for position controller, we don't concern about attitude controller etc.  We use MeasurementTypeTof, MeasurementTypeFlow, MeasurementTypeGyroscope, MeasurementTypeAcceleration. 

We use MPU9250,VL53L1X and PMW3901 sensors.we get values from sensors and want to send this values to kalman filter. We will use kalman filters outputs as position controller inputs

best regards.

**Reply by krichardsson** [reply]
> We understand that updateQueuedMeasurements() function update coredata struct. There is a switch control idiom. How this idiom work?

Measurements are popped from the queue and pushed to the right measurement model for further processing and update of the filter.

> Which case is necesssary for position controller? Our main purpose is prepare inputs for position controller, we don't concern about attitude controller etc.

I'm not sure I fully understand the questions, the cases are related to the type of measurement rather the the state in the filter.
The state contains position, velocity and attitude error, they will all be updated when you push measurements into it.
